src/Orleans/Streams/Internal/StreamDirectory.cs

```diff
@@ -20,7 +20,13 @@ internal StreamDirectory()
 
         internal IAsyncStream<T> GetOrAddStream<T>(StreamId streamId, Func<IAsyncStream<T>> streamCreator)
         {
-            return allStreams.GetOrAdd(streamId, _ => streamCreator()) as IAsyncStream<T>;
+            var stream = allStreams.GetOrAdd(streamId, _ => streamCreator()) as IAsyncStream<T>;
+            if (stream == null)
+            {
+                throw new Runtime.OrleansException($"Stream type mismatch. Please use GetStream<T>(Guid id, string streamNamespace) with the correct type, which is the type application first referenced the stream with.");
```

Comment:
Just my opinion here, but I'm not sure we should be making suggestions, rather than just telling the user what is wrong.  A stream can only support a single type of data.  The generic type of the stream requested does not match the previously requested type.  If we can get at the previously requested type to report the conflicting types, that would be great, but if that's hard to get, that's ok as it's not strictly necessary.